src/core/csv/reader_parsers.cc

```diff
@@ -652,8 +687,8 @@ void parse_string(FreadTokenizer& ctx) {
           // if there is a ", afterwards but before the next \n, use that; the field was quoted and it's still case (i) above.
           // Otherwise break here at this first sep as it's case (ii) above (the data contains a quote at the start and no sep)
           ch2 = ch;
-          while (*++ch2 && *ch2!='\n' && *ch2!='\r') {
-            if (*ch2==quote && (ch2[1]==sep || ch2[1]=='\r' || ch2[1]=='\n')) {
+          while (ch < ctx.eof && *++ch2 && *ch2!='\n' && *ch2!='\r') {
+            if (ch + 1 < ctx.eof && *ch2==quote && (ch2[1]==sep || ch2[1]=='\r' || ch2[1]=='\n')) {
```

Comment:
- the eof check should use `ch2` variable instead of `ch`;
- replace pre-increment `++ch2` with an increment within the body of the while loop;
- i believe the `*++ch2` condition (which means `*++ch2!='\0') was an old-style eof check, so should be removed

Reply:
Yes, this is a typo, here `ch2` should be checked, of course.

Reply:
As for the increment, I guess the same is true for line 679, but to replace it with an increment one should also take care about the initial value.